src/firecracker/tests/verify_dependencies.rs

```diff
@@ -0,0 +1,76 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
+use std::path::Path;
+
+use cargo_toml::{Dependency, DependencyDetail, DepsSet, Manifest};
+
+#[test]
+fn test_no_comparison_requirements() {
+    // HashMap mapping crate -> [(violating dependency, specified version)]
+    let mut violating_dependencies = HashMap::new();
+
+    let src_firecracker_path = match std::env::var("CARGO_MANIFEST_DIR") {
+        Ok(path) => path,
+        Err(_) => return, /* when running under kcov, this variable is not set. As we do not
+                           * actually run any firecracker code, we can just skip this test
+                           * without affecting coverage */
+    };
+    let src_path = format!("{}/..", src_firecracker_path);
+
+    for fc_crate in std::fs::read_dir(src_path).unwrap() {
+        let fc_crate = fc_crate.unwrap();
+        if fc_crate.metadata().unwrap().is_dir() {
+            let violating_in_crate =
+                violating_dependencies_of_cargo_toml(fc_crate.path().join("Cargo.toml"));
+
+            if !violating_in_crate.is_empty() {
+                violating_dependencies.insert(
+                    fc_crate.file_name().into_string().unwrap(),
+                    violating_in_crate,
+                );
+            }
+        }
+    }
+
+    assert_eq!(
+        violating_dependencies,
+        HashMap::new(),
+        "Dependencies should not be specified as comparison requirements. \
+        They should use caret requirements. See: \
+        https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html"
+    );
+}
+
+/// Parses the specified Cargo.toml file and returns any dependencies specified using a comparison
+/// requirements.
+///
+/// The return value maps the name of violating dependencies to the specified version
+fn violating_dependencies_of_cargo_toml(path: impl AsRef<Path>) -> HashMap<String, String> {
+    let manifest = Manifest::from_path(path).unwrap();
+
+    violating_dependencies_of_depsset(manifest.dependencies)
+        .chain(violating_dependencies_of_depsset(manifest.dev_dependencies))
+        .chain(violating_dependencies_of_depsset(
+            manifest.build_dependencies,
+        ))
+        .collect()
+}
+
+/// Returns an iterator over all dependencies in the given DepsSet specified using comparison
+/// requirements
+///
+/// The iterator produces tuples of the form (violating dependency, specified version)
+fn violating_dependencies_of_depsset(depsset: DepsSet) -> impl Iterator<Item = (String, String)> {
+    depsset.into_iter().filter_map(|(name, dependency)| {
+        match dependency {
+            Dependency::Simple(version) // dependencies specified as `libc = "0.2.117"`
+            | Dependency::Detailed(DependencyDetail {  // dependencies specified as `libc = {version = "0.2.117",...}
+                version: Some(version),
+                ..
+            }) if version.starts_with('>') => Some((name, version)),
```

Comment:
I think there are valid reasons to for example not specify a patch version (e.g. `libc = 0.2`), or to specify that we do not want to upgrade _past_ a specific patch version (e.g. `libc = "0.2, <0.2.118"`). Especially for crates that are past version 1.0, caret requirements allow to automatically drag in minor version upgrades (e.g. "1.2.3" means ">=1.2.3, <2.0.0", and to prohibit that, we would need `1.2.*` or similar. 

To cover all of these cases with a regex, if we really only want to disallow the comparison requirement, seems unergonomic to me. Besides, we do not have control over our dependencies' versioning, so we actually _need_ these additional tools in case they ever do not adhere to semver.

Reply:
I think these reasons only apply in the case we depend on crates which do not adhere to SemVer, this is something we should actively try to avoid.

In the case it is unavoidable I think we should use the exact version e.g. `libc = "=0.2.117"`.

If we think this is a likely we can adjust the regex to be `Regex::new(r"(=|)[0-9]*\.[0-9]*\.[0-9]*")` to allow either SemVer versioning or pinning to an exact version.

Reply:
Talking with jcawl@ offline, I agree that my solution is insufficient for ensuring proper dependency management, as one could still specify something as `<2340382` (ridiculously large number that will never be reached) or using wildcards. Instead of trying to catch all of these cases on a deny list, I will simply go ahead with the regex posted above. This is overly restrictive, but in praxis I have not seen use for ways to specify dependency versions beyond this. If we ever find a need to allow more requirement-types, this is always an easy change to make.